packages/shared/src/components/buttons/ButtonV2.tsx

```diff
@@ -0,0 +1,150 @@
+  Button = 'button',
+  Link = 'link',
+}
+
+export enum ButtonVariant {
+  Primary = 'primary',
+  Secondary = 'secondary',
+  Tertiary = 'tertiary',
+}
+
+export enum ButtonIconPosition {
+  Left = 'left',
+  Right = 'right',
+}
+interface CommonButtonProps {
+  kind?: ButtonKind;
+  size?: ButtonSize;
+  loading?: boolean;
+  pressed?: boolean;
+  disabled?: boolean;
+  children?: ReactNode;
+}
+
+// when color is present, variant is required
+type ColorButtonProps =
+  | { color: ButtonColor; variant: ButtonVariant }
+  | { color?: never; variant?: ButtonVariant };
+
+// when iconPosition is present, icon is required
+type IconButtonProps =
+  | { iconPosition: ButtonIconPosition; icon: IconType }
+  | { iconPosition?: never; icon?: IconType };
+
+type BaseButtonProps = CommonButtonProps & ColorButtonProps & IconButtonProps;
+
+const useGetIconWithSize = (size: ButtonSize, iconOnly: boolean) => {
+  return (icon: React.ReactElement<IconProps>) =>
+    React.cloneElement(icon, {
+      size: icon.props?.size ?? buttonSizeToIconSize[size],
+      className: classNames(icon.props.className, !iconOnly && 'icon'),
+    });
+};
+
+export type AllowedTags = keyof Pick<JSX.IntrinsicElements, 'a' | 'button'>;
+export type AllowedElements = HTMLButtonElement | HTMLAnchorElement;
+export type ButtonElementType<Tag extends AllowedTags> = Tag extends 'a'
+  ? HTMLAnchorElement
+  : HTMLButtonElement;
+
+export type ButtonProps<Tag extends AllowedTags> = BaseButtonProps &
+  HTMLAttributes<Tag> &
+  JSX.IntrinsicElements[Tag] & {
+    ref?: Ref<ButtonElementType<Tag>>;
+  };
+
+function ButtonComponent<T extends AllowedTags>(
+  {
+    kind = ButtonKind.Button,
+    variant,
+    size = ButtonSize.Medium,
+    color,
+    className,
+    icon,
+    iconPosition = ButtonIconPosition.Left,
+    loading,
+    pressed,
+    children,
+    onClick,
+    ...props
+  }: ButtonProps<T>,
+  ref?: Ref<ButtonElementType<T>>,
+): ReactElement {
+  const iconOnly = icon && !children;
+  const getIconWithSize = useGetIconWithSize(size, iconOnly);
+  const isAnchor = kind === ButtonKind.Link;
+  const Tag = isAnchor ? 'a' : 'button';
+
+  return (
+    <Tag
+      {...props}
+      {...(isAnchor ? combinedClicks(onClick) : { onClick })}
+      aria-busy={loading}
+      aria-pressed={pressed}
+      ref={ref}
+      className={classNames(
+        { iconOnly },
+        size,
+        'btn flex flex-row items-center justify-center border typo-callout font-bold no-underline shadow-none cursor-pointer select-none focus-outline relative',
+        {
+          [`btn-${variant}`]: variant,
+          [`btn-${variant}-${color}`]: color && variant,
+        },
+        className,
+      )}
+    >
+      {icon &&
+        iconPosition === ButtonIconPosition.Left &&
+        getIconWithSize(icon)}
+      {children && <span>{children}</span>}
+      {icon &&
+        iconPosition === ButtonIconPosition.Right &&
+        getIconWithSize(icon)}
+      {loading && (
```

Comment:
Can you check the behavior of our loading state here whether it renders the content still or just the spinner when it is true.

Reply:
It renders just the spinner, which is the same behavior as old button. Do we want to change it?